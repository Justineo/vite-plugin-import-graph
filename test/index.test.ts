import { execSync } from 'node:child_process'
import { existsSync, readFileSync, rmSync } from 'node:fs'
import { join } from 'node:path'
import { fileURLToPath, URL } from 'node:url'
import { build } from 'vite'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import importGraphPlugin from '../src/index'

const resolvePath = (path: string) => fileURLToPath(new URL(path, import.meta.url))

const projectRoot = resolvePath('fixtures/foo')
const outDir = join(projectRoot, 'dist')

beforeAll(() => {
  execSync('pnpm install', { cwd: projectRoot, stdio: 'inherit' })
})

afterAll(() => {
  rmSync(join(projectRoot, 'node_modules'), { recursive: true, force: true })
  rmSync(join(projectRoot, 'package-lock.json'), { force: true })
})

interface TestCase {
  description: string
  pluginOptions: Record<string, any>
  outputPath: string
  cleanupPath?: string
  customAssertion?: (graphJSON: string) => void
}

// Define your test cases
const testCases: TestCase[] = [
  {
    description: 'should handle default options correctly',
    pluginOptions: {},
    outputPath: join(projectRoot, 'import-graph.json'),
  },
  {
    description: 'should handle `absoluteModuleIds` correctly',
    pluginOptions: { absoluteModuleIds: true },
    outputPath: join(projectRoot, 'import-graph.json'),
    customAssertion: (graphJSON: string) => {
      const graph = JSON.parse(graphJSON)
      const indexKey = Object.keys(graph).find(id => id.endsWith('/index.html'))
      expect(indexKey).to.equal(join(projectRoot, 'index.html'))
    },
  },
  {
    description: 'should handle `filename` correctly',
    pluginOptions: { filename: 'foo/graph.json' },
    outputPath: join(projectRoot, 'foo/graph.json'),
    cleanupPath: join(projectRoot, 'foo'),
  },
  {
    description: 'should handle `usePrefix` correctly',
    pluginOptions: { usePrefix: true },
    outputPath: join(projectRoot, 'import-graph.json'),
  },
]

describe('basic functionalities', () => {
  for (const { description, pluginOptions, outputPath, cleanupPath, customAssertion } of testCases) {
    it(description, async () => {
      rmSync(cleanupPath || outputPath, { recursive: true, force: true })

      await build({
        root: projectRoot,
        build: {
          outDir,
          rollupOptions: {
            plugins: [importGraphPlugin(pluginOptions)],
          },
        },
      })

      expect(existsSync(outputPath)).toEqual(true)
      const graphJSON = readFileSync(outputPath, 'utf-8')

      if (!customAssertion) {
        expect(graphJSON).toMatchSnapshot()
      }
      else {
        customAssertion(graphJSON)
      }

      rmSync(cleanupPath || outputPath, { recursive: true, force: true })
      rmSync(outDir, { recursive: true, force: true })
    })
  }
})
