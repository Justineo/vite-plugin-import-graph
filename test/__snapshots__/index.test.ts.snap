// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic functionalities > should handle \`filename\` correctly 1`] = `
"{
  "\\u0000vite/modulepreload-polyfill.js": [],
  "\\u0000vite/preload-helper.js": [],
  "/css/style.css": [],
  "/index.html": [
    "\\u0000vite/modulepreload-polyfill.js",
    "/css/style.css",
    "/index.js"
  ],
  "/index.js": [
    "\\u0000vite/preload-helper.js",
    "/update.js",
    "@lukeed/uuid",
    "nanoid"
  ],
  "/update.js": [],
  "@lukeed/uuid": [],
  "nanoid": []
}"
`;

exports[`basic functionalities > should handle \`usePrefix\` correctly 1`] = `
"{
  "/css/style.css": [],
  "/index.html": [
    "/css/style.css",
    "/index.js",
    "virtual:vite/modulepreload-polyfill.js"
  ],
  "/index.js": [
    "/update.js",
    "npm:@lukeed/uuid",
    "npm:nanoid",
    "virtual:vite/preload-helper.js"
  ],
  "/update.js": [],
  "npm:@lukeed/uuid": [],
  "npm:nanoid": [],
  "virtual:vite/modulepreload-polyfill.js": [],
  "virtual:vite/preload-helper.js": []
}"
`;

exports[`basic functionalities > should handle default options correctly 1`] = `
"{
  "\\u0000vite/modulepreload-polyfill.js": [],
  "\\u0000vite/preload-helper.js": [],
  "/css/style.css": [],
  "/index.html": [
    "\\u0000vite/modulepreload-polyfill.js",
    "/css/style.css",
    "/index.js"
  ],
  "/index.js": [
    "\\u0000vite/preload-helper.js",
    "/update.js",
    "@lukeed/uuid",
    "nanoid"
  ],
  "/update.js": [],
  "@lukeed/uuid": [],
  "nanoid": []
}"
`;
